### DISCUSSION TITLE
children in Syntax Highlighter showing as object Object

### DISCUSSION THREAD

**Discussion by kartikm7** [remarkjs/remark#1337] [Q&A] [answered]
Hey guys I have been working on an Vite + Electron project, and trying to implement code block support in the chat section and I picked up the code from the docs and the String(children) returns object Object and I don't know what to do... Can some one help me out?
I suspect it could be something with my dev environment, but I am not too sure.
Also, just extra context I am working on an llm opensource chat project so the response comes from the llm in the markdown format...

#### Text on frontend
```
,[object Object],
,[object Object], ,[object Object],(,[object Object],) {
    ,[object Object], (n === ,[object Object],) ,[object Object], ,[object Object],;
    ,[object Object], {
        ,[object Object], result = n * ,[object Object],(n - ,[object Object],);
        ,[object Object], result;
    }
}

,[object Object],.,[object Object],(,[object Object],(,[object Object],));

```
The image of the console.log of the value
![image](https://github.com/remarkjs/remark/assets/108652656/96e82920-6cc1-4900-86f3-821a62835d73)

#### Actual code implementation

```            
            <Card key={index}>
              <Markdown
                className="markdown"
                remarkPlugins={[[remarkGfm, { singleTilde: false }]]}
                rehypePlugins={[rehypeHighlight]}
                components={{
                  a: (props) => {
                    return (
                      <a
                        href={props.href}
                        className="bg-foreground bg-opacity-20 opacity-70 px-1 hover:opacity-100 hover:underline transition-all"
                        target="_blank"
                        rel="noreferrer"
                      >
                        {props.children?.toString()}
                      </a>
                    )
                  },
                  code(props) {
                    const myRef = useRef<SyntaxHighlighter>(null);
                    const {children, className, node, ...rest} = props
                    const match = /language-(\w+)/.exec(className || '')
                    return match ? (
                      <SyntaxHighlighter
                        {...rest}
                        PreTag="div"
                        language={match[1]}
                        style={dark}
                        ref={myRef}
                      >{String(children).replace(/\n$/, '')}</SyntaxHighlighter>
                    ) : (
                      <code {...rest} className={className}>
                        {children}
                      </code>
                    )
                  }
                }}
              >
                {String(val.content)}
              </Markdown>
            </Card>            ```

**Comment by ChristianMurphy** [comment]
Welcome @kartikm7! 👋 
Sorry you ran into some confusion.

Could you share a runnable example of what you're seeing, ideally in a sandbox?
I tested the example from the readme it works https://stackblitz.com/edit/github-rvztvg-r6zv1s?file=src%2Fapp.tsx

I see a number of modifications in your code like tailwind and most custom elements which could be related.
I could also be what content you are passing in, have you checked what `val.content` is?

**Comment by kartikm7** [comment]
Hey @ChristianMurphy thank you so much for the prompt reply!
I am not sure how I can run the code in a sandbox since it is an electron project...  But if it's still not understandable why I will try my best to replicate it on a sandbox platform!
This is what I seen on the front end:
![image](https://github.com/remarkjs/remark/assets/108652656/36bff22c-a58c-420b-bcc8-95e8394344ab)

So val.content is the value that comes from the llm here's a log of what it does:
![image](https://github.com/remarkjs/remark/assets/108652656/189c072a-f2ee-474c-adc2-50ae86929b84)

I am using custom components in react markdown just for links and codeblocks(although the problem I face is here).
I did test the native html code tag and that parses it correctly.. My use case for it is actually is to make a codeblock component with a copy to clipboard functionality like that of gpt and others....
I don't know if this helps, but the children is of this type natively, 
> const children: React.ReactNode
I think that's why we do the String() operation, but it returns object Object for some reason

Also thought I'd share my devDependencies: Could it be the vite version?
![image](https://github.com/remarkjs/remark/assets/108652656/0a232347-bf62-4910-91cb-de42c5cbab39)


Thanks in advance man, thank you for giving your time to my query!

**Comment by kartikm7** [comment]
Just found a solution!
Had to extract just the string from children:ReactNode , found a lightweight npm package for that :
> import reactNodeToString from 'react-node-to-string'

this got it working for me!

```
              <Markdown
                className="markdown"
                rehypePlugins={[rehypeHighlight]}
                remarkPlugins={[remarkGfm]}
                components={{
                  a: (props) => {
                    return (
                      <a
                        href={props.href}
                        className="bg-foreground bg-opacity-20 opacity-70 px-1 hover:opacity-100 hover:underline transition-all"
                        target="_blank"
                        rel="noreferrer"
                      >
                        {props.children}
                      </a>
                    )
                  },
                  code(props) {
                    const myRef = useRef<SyntaxHighlighter>(null)
                    // eslint-disable-next-line @typescript-eslint/no-unused-vars
                    const { children, className, node, ...rest } = props
                    console.log(children)

                    const match = /language-(\w+)/.exec(className || '')
                    return match ? (
                      <Code language={match[1]} ref={myRef}>
                        {reactNodeToString(children).trim().replace(/\n$/, '')}
                      </Code>
                    ) : (
                      <code {...rest} className={className}>
                        {children}
                      </code>
                    )
                  }
                }}
              >
                {val.content}
              </Markdown>
              ```

**Reply by ChristianMurphy** [reply]
I'm still a bit unclear on what in your environment requires a custom serializer, but I'm glad you found a solution that works for you!

**Comment by kartikm7** [comment, marked as answer]
Hi @ChristianMurphy ! I actually just figured out why I needed to serialize, this was because I was using both `rehype-highlight` and `react-syntax-highlighter`. So this is what I think must be happening, `rehype-highlight` would pre-process the code block and convert some of the string to React Nodes and when I'd convert the `children` prop of the code-block to String and drill the prop to my custom Code component(which uses `react-syntax-highlighter)` that would cause `react-syntax-highlighter` to get a string with `[object Object]`. Removing `rehype-highlight` fixed this issue for me.

Thank you so much for being kind, and trying to help me out here!
